### ISSUE TITLE
Big slowness displaying some long pages

### ISSUE THREAD

**Issue by briceparmentier** [unclebob/fitnesse#1347]
Hi everyone,

We are using FitNesse since a long time now, and I'm taking the opportunity of this endyear period to review our database and try to do some cleanup.

We were using version 20191110, and I upgraded it to latest 20211030, hoping this would fix the issue (or at least improve our situation).

We are using almost always the !include feature that allows us to define what we call "building blocks", which are parts reusable everywhere. We thought this would reduce single tests size.

Unfortunately, we have a couple of tests that are pretty long (around 480 lines of titles and includes, file being around 15KB), and loading those ones is veeeeeery long. Could take several minutes just to display them, running FitNesse locally, not on a remote server.

Why is FitNesse so long just to render a page like this ? Is it because our massive usage of includes ?
Is there anything we can do to improve that ?
Could it be something to be improved on FitNesse source code side ?

Our main issue around this, except the fact that we have to wait a very long time in front of it when we want to edit the page, is that when running our full tests suite when we got a release is that we use triggers from continuous integration, and we can see that the first is being started around 15 minutes later!

Let me know if I can send any usefull information that could help investigate.

Thanks a lot for reading my words.

Happy new year to all !

**Comment by fhoeben** [comment]
I have no experience with render times that long, but it sounds like you are generating very big pages indeed (when all includes are loaded).

There probably is something that could be improved inside FitNesse, but I wouldn't have an idea where to start right away. Being more specific in which 'building blocks' are included in which pages can probably make a huge difference but I suppose that is not that easy to achieve. 
It could also be that inside one of your pages there is a specific feature which is causing the time to be excessive (we encountered such a feature in the usage of symbols some years ago).

Have you tried increasing the amount of (heap) memory allocated to java when starting FitNesse? Pages are loaded in memory by FitNesse and running low on usable (free) memory is a likely suspect when pages grow big.

But all in all it is hard to say something in general, without a specific page set to reproduce the issue.

Can you give some more details, and ideally some pages to reproduce the issue:
 * how do you start the wiki server (which command)?
 * do you have any plugins installed (which)?
 * what Java version are you on?
 * what kind of machine are you using (operating system, memory)?
 * how do you start the tests on the continuous integration server?

Are your tests using 'slim' or 'fit'? If you are using the slim test framework and many of you included pages define 'scenarios' for reusable blocks a lot could be gained by having scenario libraries at different levels in your test suite to define 'scopes' where scenarios are included, thereby ensuring less scenarios are loaded for each page. But figuring out which scenarios are needed at which level can be hard by hand. I wrote a tool ('[slim coverage](https://github.com/fhoeben/hsac-fitnesse-plugin)') that does so automatically. It helped me in the past to define an 'optimal' distribution of scenarios over a large test suite. In that case the load times were not as big as you report, but I did see a significant drop in page render and test start times. If you think that could help you I could provide some more details...

**Comment by briceparmentier** [comment]
Thanks for this very quick and very detailed answer! That is much appreciated!

We are running our FitNesse main server on Windows 10 on a developer desktop (which acts as a common server today for our team):
![image](https://user-images.githubusercontent.com/10654140/147823131-36282561-b145-4897-9d15-809fab5e0aec.png)
And running them locally on our desktops (which are very similar in resources) is the same.
We are running Java 8 (jre1.8 to almost latest version). I know I need to figure out how to migrate to Java 11, but I've seen a lot of changes including the fact that Java (or Oracle) have stopped delivering the JDK, and only providing the JRE. But most of Java applications are requiring JDK, so I don't know how to deal with that. As I don't really understand this domain I've let it aside for a while but now I think I'm in front of the wall ^^

On FitNesse side, we have no plugin installed (tbh I didn't know we could install plugins on FitNesse).
We are using CSlim as our fixtures are written in C++. But this point is historical and I really want to move forward and get rid of our current source code infrastructure which comes from another project we forked from, which is way too complex for our need. I tried couple of times to use CSlim library from scratch but again this is a domain I'm not expert in so I struggled writing my own fixtures in the original context of CSlim. Second point I need to go deeper into 😋 

Our FitNesse server is started from a Windows service, which launches a batch script, which calls a Python script that wraps the FitNesse server, which runs a subprocess in a thread by running the command line "java -jar PATH/TO/JAR ... blablabla".

Our continuous integration server is Jenkins, it triggers a call to the FitNesse REST API through the dedicated plugin, which retrieves JUnit results and so on.

I wish I could share some examples but unfortunately this is internal source code, and even if there is no big value in this area, we have confidentiality rules that prevent us to share it.

So first point, from your recommandations, I will try to add the Java options to increase heap amount for the FitNesse process.
Then I will do my best at trying to create a fake page that has the same behavior as our true ones we are running, whitout all confidential items.

Thanks again and have a good day, and a good party tonight ;)

**Comment by briceparmentier** [comment]
Another point I don't understand concerning the slowness loading the page, is that it is also happening when I want to show the edit page of the same script. Loading the content of the same file with Notepad++ is done almost immediately.
I can understand that rendering the page with all the includes could be longer than displaying the raw text, but not for the edit side...

**Comment by fhoeben** [comment]
I also have no idea why rendering the edit page would be that slow. It really surprises me too.

Upgrading to Java 11 should pose no problems for FitNesse, and since your tests don't use Java they should not be affected either. For FitNesse you just need the JRE (no compilation needed). I would recommend going to one of the open source distributions as Oracle's license terms are not very nice... 

Hopefully adding some more heap will improve the situation somewhat.

**Comment by jediwhale** [comment]
It's likely a problem in the parser. In the past ,we've seem rare combinations of characters near the end of long pages that cause the parser to slow down.  If you can post the exact text of the problem page, I can see if this is the cause.

**Comment by jediwhale** [comment]
I just noticed your comment about confidentiality rules.  If you can't share the problem pages, can you create a version that exhibits the problem, with all proprietary information removed? Or you can try to narrow down the cause: start with an empty page and add pieces of the problem content one by one, until the problem occurs. Then remove the last addition and break it into smaller pieces and add one by one, etc. See if you can narrow it down to a small piece of content. Sometimes it might be a typo in your content markup that's confusing the parser.  Or it may be test data that has unusual characters - try enclosing it in literal markup !- and -!.

Hope this helps.

**Comment by briceparmentier** [comment]
Thanks @jediwhale for your comments!
For now I've stopped the work so I haven't got any pieces available.
But from what I remember when we started seeing the issue (we investigate a bit and tried some different ideas like the special characters you are mentionning), it was after overpassing a given number of lines (don't remember exactly).
Next week I'll try to give more details ;)
Happy year ending to all ;)

**Comment by briceparmentier** [comment]
Hi there,
Juste a little piece of information about how our FitNesse server is launched, looking at the command line used in the task manager:
java -Xmx1024m -Dslim.port=XXXXX -Dslim.pool.size=XXXX -Dslim.debug.timeout=60 -jar .\path\to\jar\fitnesse-standalone.jar -p XXXX -e 0 -o
Does this bring any clue?

**Comment by briceparmentier** [comment]
Hello here,

We are back on the subject thanks to the arrival of a new team member.
We found that most of the waiting time was due to us not closing collapsible blocks (which starts with !*> or !**>). We were missing the closing tag *!.
In one of the tests we still have some seconds to load the page, but we went down from around 1 minute loading to 3-4 seconds!
Related issue was opened here : #1371 

This one can then be closed :)
Thanks !
